llvm-general/test/LLVM/General/Test/Linking.hs

```diff
@@ -59,9 +59,9 @@ tests = testGroup "Linking" [
         ]      
 
     Module { moduleDefinitions = defs } <- withContext $ \context -> 
-      withModuleFromAST' context ast0 $ \m0 ->
-        withModuleFromAST' context ast1 $ \m1 -> do
-          runExceptT $ linkModules False m0 m1
-          moduleAST m0
+      withModuleFromAST' context ast0 $ \dest -> do
+        src <- failInIO $ astToFFIModule context ast1
+        failInIO $ linkModules dest src
```

Comment:
This test demonstrates my concern. Now users who want to link have to make their "src" module one way, but users who don't or users who do but are making their "dest" should make them another way, and all users have to be mindful of the difference.

A better type for linkModules would express its constraints better. The previous behavior was a compromise. Is the change necessary for 3.9?